core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -743,6 +744,9 @@ public PlanWithProperties visitExchange(ExchangeNode node, StreamPreferredProper
                         any().withOrderSensitivity(),
                         any().withOrderSensitivity());
             }
+            if (isForceFixedDistributionForPartitionedOutputOperatorEnabled(session) && node.isHashPartitionedExchange()) {
```

Comment:
Shouldn't this behavior only be induced for partitioned output operators on top of table scans? Intermediate stages shouldn't have the same flushing behaviors, so condition seems potentially overly-broad, but maybe that's handled by `planAndEnforceChildren` somehow?